ops/pebble.py

```diff
@@ -1976,3 +1972,232 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+        if content_disposition != 'form-data':
+            raise ProtocolError(
+                'unexpected content disposition: {!r}'.format(content_disposition))
+
+        name = self._headers.get_param('name', header='content-disposition')
+        if name == 'files':
+            filename = self._headers.get_filename()
+            if filename is None:
+                raise ProtocolError('multipart "files" part missing filename')
+            self._prepare_tempfile(filename)
+        elif name != 'response':
+            raise ProtocolError(
+                'unexpected name in content-disposition header: {!r}'.format(name))
+
+        self._part_type = name
+
+    def _process_body(self, data: bytes, done=False):
+        if self._part_type == 'response':
+            self._response_data.extend(data)
+            if done:
+                if len(self._response_data) > self._max_lookahead:
+                    raise ProtocolError('response end marker not found')
+                self._response = json.loads(self._response_data.decode())
+                self._response_data = bytearray()
+        elif self._part_type == 'files':
+            if done:
+                # This is the final write.
+                outfile = self._get_open_tempfile()
+                outfile.write(data)
+                outfile.close()
+                self._headers = None
+            else:
+                # Not the end of file data yet. Don't open/close file for intermediate writes
+                outfile = self._get_open_tempfile()
+                outfile.write(data)
+
+    def remove_files(self):
+        """Remove all temporary files on disk."""
+        for file in self._files.values():
+            os.unlink(file.name)
+        self._files.clear()
+
+    def feed(self, data: bytes):
+        """Provide more data to the running parser."""
+        self._parser(data)
+
+    def _prepare_tempfile(self, filename):
+        tf = tempfile.NamedTemporaryFile(delete=False)
+        self._files[filename] = tf
+        self.current_filename = filename
+
+    def _get_open_tempfile(self):
+        return self._files[self.current_filename]
+
+    def get_response(self):
+        """Return the deserialized JSON object from the multipart "response" field."""
+        return self._response
+
+    def filenames(self):
+        """Return a list of filenames from the "files" parts of the response."""
+        return list(self._files.keys())
+
+    def get_file(self, path, encoding):
+        """Return an open file object containing the data."""
+        mode = 'r' if encoding else 'rb'
+        # We're using text-based file I/O purely for file encoding purposes, not for
+        # newline normalization.  newline='' serves the line endings as-is.
+        newline = '' if encoding else None
+        return open(self._files[path].name, mode, encoding=encoding, newline=newline)
+
+
+def _multipart_parser(
```

Comment:
Is this done as a separate function to make unit testing easier? I suppose that's fair enough, however, it seems to me it'd be simpler to just make these methods part of the main `_FilesParser` class, to avoid having to pass in the `handle_*` functions. It would just call `self.handle_*` and update its own state. That said, I don't feel strongly about this if the separation makes testing easier (assuming we're also testing the rest of the `_FilesParser` methods).